rich-text: set up autogenerated API docs

oandregal opened the pull request:
**Related to** https://github.com/WordPress/gutenberg/issues/14227

This PR sets up auto-generated API docs for the `rich-text` package through the `npm run docs:build` command and updates its README file.

## Notable modifications

- Fixed optional parameters. Should be `{type} [param]`, not `{?type} param`.
- Fixed returns when type can be undefined. Should be `{type|undefined}`, not `{?type}`.
- New symbols added that were undocumented:
  - `insertObject`
  - `unregisterFormatType`
- Symbols that were documented and are removed in this PR: `apply`.
- Symbols that stay undocumented:
  - `changeListType`
  - `charAt`
  - `getSelectionEnd`
  - `getSelectionStart`
  - `indentListItems`
  - `insertLineBreak`
  - `insertLineSeparator`
  - `isEmptyLine`
  - `LINE_SEPARATOR`
  - `outdentListItems`
  - `unstableToDom`

## Test instructions

- Delete or modify the contents at `packages/rich-text/README.md` within the start `<!-- START TOKEN(Autogenerated API docs) -->` and end `<!-- END TOKEN(Autogenerated API docs) -->` tokens.
- Execute `npm run docs:build` and verify they are re-created.
- Review that the API docs match the JSDoc of the exported symbols.

oandregal committed:
Autogenerated docs and setup for rich-text

oandregal committed:
Use a central script instead of one per package

oandregal committed:
Prepare to add more packages

oandregal committed:
Use bin script instead of npm script

oandregal committed:
Update JSDoc from old README and regenerate

oandregal committed:
Do not document unstable APIs

ellatrix commented:
Thanks! I would rather not document any functions that are currently not documented in `master`. I guess either we should remove the functions or mark them as unstable (e.g. all the list specific functions).

ellatrix commented on packages/rich-text/README.md:
```diff
@@ -14,169 +14,507 @@ _This package assumes that your code will run in an **ES2015+** environment. If
+-   **$1.current** `HTMLElement`: The live root node to apply the element tree to.
+-   **$1.multilineTag** `?string`: Multiline tag.
+-   **$1.multilineWrapperTags** `?Array`: Tags where lines can be found if nesting is possible.
+-   **$1.prepareEditableTree** `?Array`: Functions to prepare the editable tree.
```
This attribute should be marked as unstable. It's was added for the annotations API, which is also unstable.

nosolosw replied:
Removing the att from the JSDoc will remove it from this as well. I just did that cfae06862fb708a367aa950dcb4d9573a637a220

This was something I documented because it wasn't (I didn't know it was supposed to be unstable/private).

ellatrix commented on packages/rich-text/README.md:
````diff
@@ -14,169 +14,507 @@ _This package assumes that your code will run in an **ES2015+** environment. If
-```js
-isEmpty( value: Object ): boolean
-```
+### changeListType
````
All list related functions should be marked unstable. I should have done this when adding it but I forgot. :(

nosolosw replied:
Added `changeListType` to the ignore list at 3d61f40

ellatrix commented on packages/rich-text/README.md:
```diff
@@ -14,169 +14,507 @@ _This package assumes that your code will run in an **ES2015+** environment. If
+
+`?Object` Active format object of the specified type, or undefined.
+
+### getSelectionEnd
```
`value.start` and `value.end` are going to stay, so I guess these are a bit useless. Maybe we should remove them? They were never documented before.

nosolosw replied:
Not sure I follow how this comment relates to `getSelectionEnd` :) FWIW, I'm barely familiar with the `rich-text` package.

ellatrix commented on packages/rich-text/README.md:
````diff
@@ -14,169 +14,507 @@ _This package assumes that your code will run in an **ES2015+** environment. If
-	?multilineWrapperTags: Array,
-} ): Object
-```
+### apply
````
This one should ideally also be marked unstable, as it's DOM-specific. It was exported so that the `RichText` component in the editor package could make use of it, but I consider this an internal function.

nosolosw replied:
Added `apply` to the ignore list at 3d61f40

gziolo commented:
> I guess either we should remove the functions or mark them as unstable (e.g. all the list specific functions).

Ideally, we fix export statement and JSDocs so the tool could do its job :)

ellatrix commented:
> Ideally, we fix export statement and JSDocs so the tool could do its job :)

It would ignore function names prefixed with `unstable` right?

oandregal committed:
Merge docs:build and docs:generate

oandregal committed:
Remove unstable param

oandregal committed:
Ignore apply and changeListType symbols

nosolosw commented:
@ellatrix Added `apply` and `changeListType` to the list of symbols to be ignored. Let me know if there are more we should ignore.

Although this works it makes me think about how `apply` and `changeListType` can communicate they are for internal use only while not having a `unstable`/`experimental`/etc prefix. One thing we can do is to add the [`@private` JSDoc tag](http://usejsdoc.org/tags-private.html). Thoughts?

Although the auto-documentation tool doesn't currently ignore the exported symbols that contain a `@private` tag, we could easily add it. It'll be useful for other things as well, for example, to avoid documenting symbols that are exported for testing purposes.

chrisvanpatten commented:
@nosolosw I really like the idea of supporting `@private`. Feels very appropriate.

nerrad commented:
> One thing we can do is to add the @private JSDoc tag. Thoughts?

Big ➕from me on this.

chrisvanpatten commented:
@nerrad wish I could give you a bigger 👍 on that. Agree completely!

oandregal committed:
Use new script

oandregal committed:
Update output

oandregal committed:
Refactor to use switch

ellatrix commented:
@nosolosw Agreed with the `@private` and `__unstable`. Earlier I meant all the list related function, not just `changeListType`, but actually I think it's okay to expose them. The others I've removed in #14239.

oandregal committed:
Ignore functions that are about to be removed

oandregal committed:
Update docs

nosolosw commented:
@ellatrix so, [this](https://github.com/WordPress/gutenberg/pull/14220/files#diff-8a16d6e82b08e8c6804f266d34da8597R18) the final list to ignore, is that OK? Note that I've also added the ones to be removed just so we can merge this independently of any other PR.

ellatrix commented:
@nosolosw You can remove `changeListType`. It doesn't make sense to ignore `changeListType` when not ignoring the other list functions.

oandregal committed:
Ignore also indentListItem and outdentListItems

nosolosw commented:
This is the final list of symbols that will be ignored: `apply`, `changeListType`, `chartAt`, `getSelectionEnd`, `getSelectionStart`, `indentListItems`, `insertLineBreak`, `outdentListItems`, `unstableToDom`.

This is ready for review.

ellatrix commented on packages/rich-text/src/join.js:
```diff
@@ -11,7 +11,7 @@ import { normaliseFormats } from './normalise-formats';
  *
  * @param {Array}         values    An array of values to join.
- * @param {string|Object} separator Separator string or value.
+ * @param {?(string|Object)} separator Separator string or value.
```
Hm, it looks like `separator` is required here?

nosolosw replied:
The `separator` arg is given a default value of `''` if none is provided, so I guess is optional? For reference: [current docs](https://github.com/WordPress/gutenberg/tree/master/packages/rich-text#join) also mark it as that can be `null`.

ellatrix replied:
Ah, correct.

ellatrix commented on packages/rich-text/src/to-dom.js:
```diff
@@ -164,10 +164,11 @@ export function toDom( {
+ * @param {Object}      $1 Named arguments.
+ * @param {Object}      $1.value Value to apply.
+ * @param {HTMLElement} $1.current The live root node to apply the element tree to.
+ * @param {?string}      $1.multilineTag Multiline tag.
```
It looks like this is one space too far indented.

ellatrix commented on packages/rich-text/src/to-html-string.js:
```diff
@@ -20,8 +20,8 @@ import { toTree } from './to-tree';
  * @param {Object} $1.value                Rich text value.
- * @param {string} $1.multilineTag         Multiline tag.
- * @param {Array}  $1.multilineWrapperTags Tags where lines can be found if
+ * @param {?string} $1.multilineTag         Multiline tag.
```
Here as well. Too far indented

ellatrix commented on packages/rich-text/src/split.js:
```diff
@@ -10,8 +10,8 @@ import { replace } from './replace';
  * @param {Object}        value   Value to modify.
- * @param {number|string} string  Start index, or string at which to split.
- * @param {number}        end     End index.
+ * @param {?(number|string)} string  Start index, or string at which to split.
```
Looks like the indentation is not correct here.

ellatrix commented on packages/rich-text/src/apply-format.js:
```diff
@@ -17,8 +17,8 @@ import { normaliseFormats } from './normalise-formats';
  * @param {Object} format     Format to apply.
- * @param {number} startIndex Start index.
- * @param {number} endIndex   End index.
+ * @param {?number} startIndex Start index.
```
I think anywhere there's a change in inline docs, the indentation was not adjusted. :)

oandregal committed:
Fix typo

oandregal committed:
Update docs

- Fix return undefined or type: should be "{type|undefined}", not "{?type}".
- Fix optional type: should be "{type} [param]", not "{?type} param".
- Fix list of undocumented symbols.
- Updated JSDoc from old README (added new params, etc).

ellatrix commented on packages/rich-text/src/char-at.js:
```diff
@@ -5,7 +5,7 @@
  * @param {string} index Index to use.
  *
- * @return {?string} A one character long string, or undefined.
+ * @return {string|undefined} A one character long string, or undefined.
```
Is this a new recommendation? Do we have any docs on writing inline docs?

nosolosw replied:
As per the [`@type` tag documents](http://usejsdoc.org/tags-type.html) this is one of the issues I've identified:

- we should use `?string` if the type can be `null` or a `string`.
- we should use `string|undefined` if the type can be a `string` or `undefined`.

From the text description, I understood that we want the second.

ellatrix replied:
Nice, I like that it's clearer as well.

nosolosw commented:
@ellatrix I missed tons of spaces! Sorry about that.

Initially, I updated the JSDoc comments with info that was on the [current README docs](https://github.com/WordPress/gutenberg/tree/master/packages/rich-text#join) but wasn't present in the JSDoc comment. Having to deal with a myriad of packages at https://github.com/WordPress/gutenberg/issues/14227 I didn't know better.

However, I ended up doing a full audit of these API docs. I added the notable modifications this PR introduces in the description. Would you mind taking a look at that to see if it makes sense to you?

The only change that stands out to me as potentially wrong is [the suggestion](https://github.com/WordPress/gutenberg/pull/14220#discussion_r262152140) to remove `apply` from the docs. If it was already documented should we keep it that way (I'm unfamiliar with how we proceed with this kind of thing)?

ellatrix commented on packages/rich-text/src/insert-line-separator.js:
```diff
@@ -11,9 +11,9 @@ import { LINE_SEPARATOR } from './special-characters';
- * @param {Object} value         Value to modify.
- * @param {number} startIndex    Start index.
- * @param {number} endIndex      End index.
+ * @param {Object} value        Value to modify.
```
Slight inconsistency here, compared to https://github.com/WordPress/gutenberg/pull/14220/files#diff-9ab2f40d2dac2f7335282d5df03a013fR18.

nosolosw replied:
ok, fixed spacing

ellatrix commented on packages/rich-text/src/remove.js:
```diff
@@ -9,9 +9,9 @@ import { create } from './create';
- * @param {Object} value      Value to modify.
- * @param {number} startIndex Start index.
- * @param {number} endIndex   End index.
+ * @param {Object}  value       Value to modify.
```
Indentation here is inconsistent across the updated docs.

nosolosw replied:
Fixed at 50d18e5

ellatrix commented on packages/rich-text/src/special-characters.js:
```diff
@@ -1,2 +1,5 @@
+/**
+ * Keycode for line separator.
```
This is a string containing the "line separator" character, not a keycode?

nosolosw replied:
Ah, wasn't sure what it was :)

nosolosw replied:
Fixed at 70647bc

ellatrix replied:
Thanks! :)

oandregal committed:
Fix comment

oandregal committed:
Fix spaces

oandregal committed:
Fix spaces

oandregal committed:
Ignore more symbols

oandregal committed:
fix spaces

oandregal committed:
fix spaces

oandregal committed:
fix spaces

ellatrix reviewed (approved):
Looks great, thanks!
